### ISSUE TITLE
Invalid property 'cache.hazelcast.management-center[enabled]

### ISSUE THREAD

**Issue by dapperdanman** [jhipster/generator-jhipster#7062]
##### **Overview of the issue**

I generated a Microservice Gateway and I'm trying to start it up. I get this error:

config/AsyncConfiguration.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'io.github.jhipster.config.JHipsterProperties': Could not bind properties to JHipsterProperties (prefix=jhipster, ignoreInvalidFields=false, ignoreUnknownFields=false, ignoreNestedProperties=false); nested exception is org.springframework.beans.NotWritablePropertyException: Invalid property 'cache.hazelcast.management-center[enabled]' of bean class [io.github.jhipster.config.JHipsterProperties$Cache]: Cannot access indexed value in property referenced in indexed property path 'management-center[enabled]'; nested exception is org.springframework.beans.NotReadablePropertyException: Invalid property 'cache.hazelcast.management-center[enabled]' of bean class [io.github.jhipster.config.JHipsterProperties$Cache]: Bean property 'cache.hazelcast.management-center[enabled]' is not readable or has an invalid getter method: Does the return type of the getter match the parameter type of the setter?
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:667)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:366)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1264)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553)

##### **Motivation for or Use Case**

A new gateway should be able to start!

##### **Reproduce the error**

Generate new microservice gateway.

##### **JHipster Version(s)**

4.13.3
I have gotten this to work before with v4.5.5

##### **JHipster configuration**

$ jhipster info
Using JHipster version installed locally in current project's node_modules
Executing jhipster:info
Options: 
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
catdm-force-gateway@0.0.0 /local/data/CatdmForceGateway
└── generator-jhipster@4.13.3 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "catdm.force.gateway",
      "nativeLanguage": "en",
      "microservicePath": "../CatdmForceDocsMicroService"
    },
    "jhipsterVersion": "4.13.3",
    "baseName": "CatdmForceGateway",
    "packageName": "catdm.force.gateway",
    "packageFolder": "catdm/force/gateway",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "hazelcast",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Memory",
    "prodDatabaseType": "mysql",
    "searchEngine": "elasticsearch",
    "messageBroker": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "angularX",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling",
      "cucumber",
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "it"
    ]
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_152"
Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.152-b16, mixed mode)

git version 2.1.0

node: v8.9.4

npm: 5.6.0

yeoman: 2.0.0

yarn: 1.3.2

Congratulations, JHipster execution is complete!

##### **Browsers and Operating System**

Red Hat Enterprise Linux 6

- [x ] Checking this box is mandatory (this is just to show you read everything)

**Comment by jdubois** [comment]
How do you start your gateway, just doing "./mvnw"?
I couldn't reproduce.
I don't understand why this property wouldn't exist in your JHipsterProperties, it comes from our jhipster lib, so it's not generated.

**Comment by jdubois** [comment]
I just can't reproduce... That property should be there, I don't understand why it wouldn't be there, it's not even generated, it's imported from our server lib.

Closing as I can't reproduce.

**Comment by dapperdanman** [comment]
$WARFILE --spring.profiles.active=dev,devfoo

I created a new "profile" named devfoo that points to different hostnames than localhost. I copied the prod profile and tweaked the hostnames. 

I'm not using mvnw, but my own internal Maven v3.0.4 installation in my $PATH to compile.

**Comment by jdubois** [comment]
And are you sure you have the JHipster server lib v1.3.0? It should come from the JHipster Dependencies BOM.

**Comment by dapperdanman** [comment]
I have this in my pom.xml:

<jhipster-dependencies.version>0.1.6</jhipster-dependencies.version>

and a <dependencyManagement> section that uses it with scope "import"

I added a <distributionManagement> section above this so Maven uses my internal Nexus repo.

**Comment by jdubois** [comment]
Sorry no, your config is correct.
Could you do:

    mvn dependency:tree | grep jhipster

That should give you

    io.github.jhipster:jhipster:jar:1.3.0:compile

And the JHisperProperties class comes from here

**Comment by dapperdanman** [comment]
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>io.github.jhipster</groupId>
                <artifactId>jhipster-dependencies</artifactId>
                <version>${jhipster-dependencies.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
            
        </dependencies>
    </dependencyManagement>

**Comment by dapperdanman** [comment]
`mvn dependency:tree | grep jhipster`

`[INFO] +- io.github.jhipster:jhipster:jar:1.3.0:compile`

**Comment by jdubois** [comment]
We're getting closer :-)
And in your `application-dev.yml`, do you have Hazelcast configured:

```
    cache: # Cache configuration
        hazelcast: # Hazelcast distributed cache
            time-to-live-seconds: 3600
            backup-count: 1
            management-center: # Full reference is available at: http://docs.hazelcast.org/docs/management-center/3.9/manual/html/Deploying_and_Starting.html
                enabled: false
                update-interval: 3
                url: http://localhost:8180/mancenter

**Comment by dapperdanman** [comment]
I do have that section. My application-devfoo.yml file also has that section, but the url: value is empty.

**Comment by jdubois** [comment]
The stacktrace is for `'cache.hazelcast.management-center[enabled]'` - but if you leave that property empty you would have a clean error message
Anyway for me the issue looks rather that you don't have the correct jhipster lib in your classpath - could you try running your app with Maven, without packaging it? Just run `mvn` (and also `./mvnw` as you have an old Maven version - maybe that causes the issue?)

**Comment by dapperdanman** [comment]
I did get the Gateway started using "mvn". I'm using the JHipster Registry too. The Gateway registers, but the status is "down" on the Registry status page. 

Ideally, I want these servers to run on a machine without Maven installed, so that is why I try to use shell scripts to launch them. Anyway, making progress!

**Comment by gmarziou** [comment]
> Ideally, I want these servers to run on a machine without Maven installed, so that is why I try to use shell scripts to launch them.

Of course! Nobody uses maven in production. The generated war is an executable jar that embeds a shell script, so starting it is very simple and you can easily define a systemd service unit from it.

**Comment by dapperdanman** [comment]
OK, I think I got it working. I don't want to use the maven-frontend-plugin in the prod profile to download and install yarn/etc from the Internet. I'm in a corporate environment behind a proxy. I'll manually setup my builder nodes on my workstation and Jenkins slaves to have Yarn/etc available. I modified the prod profile to use the maven-exec plugin to use my existing yarn install to run "yarn install", "yarn test...", "yarn run webpack:prod"... I also had to add "--ignore_ssl" to the "yarn install" command to work around corporate self-signed certificate issues in the proxy. 

I had commented out that stuff before, and that really broke things. :)

I think there is an opportunity here to add a new profile to the POM, or document these types of changes on the website somewhere to help other people. Not sure how you want to handle that. 

Thanks for your patience with me.
